addon/components/frost-table-header.js

```diff
@@ -41,12 +46,115 @@ export default Component.extend({
 
   // == Computed Properties ===================================================
 
+  @readOnly
+  @computed('css')
+  _categoryRowClass (css) {
+    return `${css}-categories`
+  },
+
+  @readOnly
+  @computed('css')
+  _columnRowClass (css) {
+    return `${css}-columns`
+  },
+
+  @readOnly
+  @computed('columns')
+  _hasCategories (columns) {
+    let hasCategories = false
+    columns.forEach((column) => {
+      if (column.category) {
+        hasCategories = true
+        return
+      }
+    })
+    return hasCategories
+  },
+
+  @readOnly
+  @computed('columns')
+  _categoryColumns (columns) {
+    let categoryColumns = A([])
+    let index = 0
+    columns.forEach((column) => {
+      let lastCategoryColumn = categoryColumns.get('lastObject')
+      const categoryLabel = column.category ? column.category : ''
+      if (lastCategoryColumn && lastCategoryColumn.label === categoryLabel) {
+        lastCategoryColumn.span += 1
+      } else {
+        categoryColumns.addObject({
+          index: index++,
+          label: categoryLabel,
+          span: 1,
+          className: column.categoryClassName,
+          renderer: column.headerCategoryRenderer
+        })
+      }
+    })
+    return categoryColumns
+  },
+
   // == Functions =============================================================
 
+  setupRows () {
+    const rowTag = this.get('rowTagName')
+    const selectable = this.get('isSelectable')
+    const categoryColumns = this.get('_categoryColumns')
+    const categoryRowClass = this.get('_categoryRowClass')
+    const columnRowClass = this.get('_columnRowClass')
+
+    // Wrap category and regular header columns into separate rows
+    const lastCategoryIndex = categoryColumns.length + (selectable ? 1 : 0)
```

Comment:
ah ok maybe add a function to wrap this or find a way that it's clear why we are doing this 1 : 0 business.